### PR TITLE
Convert parachain-staking for use with cumulus and pallet-session

### PR DESCRIPTION
**Pull request by nvengal**
This PR:
1. Converts parachain-staking to use default polkadot branches for substrate pallets
2. Implements the required traits for use with the session pallet and cumulus concensus

Reference implementations:
- https://github.com/bifrost-finance/parachain-staking
- https://github.com/KILTprotocol/mashnet-node/tree/develop/pallets/parachain-staking

### CODE REVIEW

File: pallets/parachain-staking/src/lib.rs [hunk 1]
```diff
@@ -1652,26 +1652,87 @@ pub mod pallet {
...
+		// Add reward points to block authors:
+		// * 20 points to the block producer for producing a block in the chain
 		fn note_author(author: T::AccountId) {
 			let now = <Round<T>>::get().current;
 			let score_plus_20 = <AwardedPts<T>>::get(now, &author).saturating_add(20);
 			<AwardedPts<T>>::insert(now, author, score_plus_20);
 			<Points<T>>::mutate(now, |x| *x = x.saturating_add(20));
 		}
+
+		fn note_uncle(_author: T::AccountId, _age: T::BlockNumber) {}
```
Comment (rhuttman) [inline pallets/parachain-staking/src/lib.rs:1671]: why is this a stub? Maybe even add a comment saying why. Same for the others down below.

File: pallets/parachain-staking/src/lib.rs [hunk 2]
```diff
@@ -1652,26 +1652,87 @@ pub mod pallet {
...
+	impl<T: Config> pallet_session::ShouldEndSession<T::BlockNumber> for Pallet<T> {
+		fn should_end_session(now: T::BlockNumber) -> bool {
+			let round = <Round<T>>::get();
+			return round.should_update(now);
+		}
+	}
+
+	impl<T: Config> frame_support::traits::EstimateNextSessionRotation<T::BlockNumber> for Pallet<T> {
+		fn average_session_length() -> T::BlockNumber {
+			T::BlockNumber::from(<Round<T>>::get().length)
```
Comment (rhuttman) [inline pallets/parachain-staking/src/lib.rs:1709]: so a round is a session?

File: pallets/parachain-staking/src/lib.rs [hunk 3]
```diff
@@ -1652,26 +1652,87 @@ pub mod pallet {
...
 		}
+
+		fn note_uncle(_author: T::AccountId, _age: T::BlockNumber) {}
 	}
 
-	impl<T: Config> nimbus_primitives::CanAuthor<T::AccountId> for Pallet<T> {
-		fn can_author(account: &T::AccountId, _slot: &u32) -> bool {
-			Self::is_selected_candidate(account)
+	impl<T: Config> pallet_session::SessionManager<T::AccountId> for Pallet<T> {
+		fn new_session(new_index: sp_staking::SessionIndex) -> Option<Vec<T::AccountId>> {
```
Comment (justinzhou93) [inline pallets/parachain-staking/src/lib.rs:1675]: confused on what these functions are used for

File: pallets/parachain-staking/src/lib.rs (hunk 2 above)
Comment (nvengal) [reply pallets/parachain-staking/src/lib.rs:1709]: For all intents and purposes, yes.
`session` is the parlance of pallet-session. A session is a period of time that has constant set of collators.
`round` is the parlance of parachain-staking. At the start of every new round a new set of collators is chosen.
So, when relating sessions to rounds we can say that a new session starts when a new round starts.

File: pallets/parachain-staking/src/lib.rs (hunk 1 above)
Comment (nvengal) [reply pallets/parachain-staking/src/lib.rs:1671]: I have to define these methods in order to satisfy the trait. In this case, the `pallet_authorship::EventHandler` trait needs to be able to handle the `note_uncle` event.
For our current implementation, we don't care about when a block is an uncle so this is a noop.

For reference, here's the compilation error if the stubbed methods are removed:
```
error[E0046]: not all trait items implemented, missing: `note_uncle`
    --> /Users/nikhil/oak/moonbeam/pallets/parachain-staking/src/lib.rs:1661:2
     |
1661 |     impl<T: Config> pallet_authorship::EventHandler<T::AccountId, T::BlockNumber> for Pallet<T> {
     |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ missing `note_uncle` in implementation
     |
     = help: implement the missing item: `fn note_uncle(_: Author, _: BlockNumber) { todo!() }`

error[E0046]: not all trait items implemented, missing: `end_session`, `start_session`
    --> /Users/nikhil/oak/moonbeam/pallets/parachain-staking/src/lib.rs:1672:2
     |
1672 |     impl<T: Config> pallet_session::SessionManager<T::AccountId> for Pallet<T> {
     |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ missing `end_session`, `start_session` in implementation
     |
     = help: implement the missing item: `fn end_session(_: u32) { todo!() }`
     = help: implement the missing item: `fn start_session(_: u32) { todo!() }`
```

Happy to add a comment if you think that it'll be useful, but I'm not sure what I'd say other than something like "we don't currently care about uncles" which seems less than useful.

File: pallets/parachain-staking/src/lib.rs (hunk 3 above)
Comment (nvengal) [reply pallets/parachain-staking/src/lib.rs:1675]: These functions satisfy traits for use with the session pallet.
See https://github.com/OAK-Foundation/OAK-blockchain/blob/dfd9523e8d9a1d41606cb90a8abe03ca7e085541/runtime/neumann/src/lib.rs#L440-L442

File: pallets/parachain-staking/src/lib.rs (hunk 1 above)
Comment (rhuttman) [reply pallets/parachain-staking/src/lib.rs:1671]: Just seems weird that we need to fulfill a trait but don't need all the fcns in it. In theory they exist for a reason

Comment (nvengal) [reply pallets/parachain-staking/src/lib.rs:1671]: If it makes you feel better, even the implementation of this trait in cumulus ignores uncles- https://github.com/paritytech/cumulus/blob/9aead492b71990aa6f7bfc0cd35108078f385ae6/pallets/collator-selection/src/lib.rs#L491-L493

Comment (nvengal) [reply pallets/parachain-staking/src/lib.rs:1671]: Same for the session manager trait- https://github.com/paritytech/cumulus/blob/9aead492b71990aa6f7bfc0cd35108078f385ae6/pallets/collator-selection/src/lib.rs#L517-L522

Comment (rhuttman) [reply pallets/parachain-staking/src/lib.rs:1671]: yes and no haha

**Review by justinzhou93 — APPROVED** [review]
lgtm - makes more sense now! thanks!

**Review by rhuttman — APPROVED** [review]

**Pull request merged**